File: solutions/dynamic-programming/070-climbing-stairs.py
"""
70. Climbing Stairs
Easy

You are climbing a staircase. It takes n steps to reach the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you
climb to the top?

Example:
Input: n = 3
Output: 3
Explanation: There are three ways to climb to the top.
1. 1 step + 1 step + 1 step
2. 1 step + 2 steps
3. 2 steps + 1 step

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
This is the classic Fibonacci problem in disguise! To reach step n, you can
either come from step (n-1) by taking 1 step, or from step (n-2) by taking 2 steps.
So: ways(n) = ways(n-1) + ways(n-2)

### PATTERN RECOGNITION:
- n=1: 1 way (1)
- n=2: 2 ways (1+1, 2)
- n=3: 3 ways (1+1+1, 1+2, 2+1)
- n=4: 5 ways (1+1+1+1, 1+1+2, 1+2+1, 2+1+1, 2+2)
- Notice: 1, 2, 3, 5... This is Fibonacci sequence!

### APPROACH:
1. Base cases: ways(1)=1, ways(2)=2
2. For any step n: ways(n) = ways(n-1) + ways(n-2)
3. Use bottom-up DP to avoid redundant calculations

### WHY FIBONACCI?
- From step (n-1): one way to reach n (take 1 step)
- From step (n-2): one way to reach n (take 2 steps)
- These are the ONLY ways to reach step n
- So total ways = ways to reach (n-1) + ways to reach (n-2)

### OPTIMIZATION:
Instead of storing all values, we only need the last two values,
making space complexity O(1) instead of O(n).

</details>
"""

class Solution:
    def climbStairs(self, n: int) -> int:
        """
        Approach: Dynamic Programming (Fibonacci)
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        if n <= 2:
            return n

        # Only need to track last two values
        prev2 = 1  # ways to reach step i-2
        prev1 = 2  # ways to reach step i-1

        for i in range(3, n + 1):
            current = prev1 + prev2
            prev2 = prev1
            prev1 = current

        return prev1

    def climbStairsDP(self, n: int) -> int:
        """
        Approach: Dynamic Programming with array
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        if n <= 2:
            return n

        dp = [0] * (n + 1)
        dp[1] = 1
        dp[2] = 2

        for i in range(3, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]

        return dp[n]

    def climbStairsRecursive(self, n: int) -> int:
        """
        Approach: Recursive with memoization
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        memo = {}

        def helper(n):
            if n <= 2:
                return n

            if n in memo:
                return memo[n]

            memo[n] = helper(n - 1) + helper(n - 2)
            return memo[n]

        return helper(n)


# Test cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1
    n1 = 2
    print(f"Input: n = {n1}")
    print(f"Output: {solution.climbStairs(n1)}")  # 2

    # Test case 2
    n2 = 3
    print(f"Input: n = {n2}")
    print(f"Output: {solution.climbStairs(n2)}")  # 3

    # Test case 3
    n3 = 5
    print(f"Input: n = {n3}")
    print(f"Output: {solution.climbStairs(n3)}")  # 8

File: solutions/dynamic-programming/alternatives/070-climbing-stairs.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 *
 * **Step 1:** [description]
 *
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Main solution for Problem 070: Climbing Stairs
 *
 * @param {number} n - Number of steps to reach the top
 * @return {number} - Number of distinct ways to climb to the top
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
function solve(n) {
    if (n <= 2) {
        return n;
    }

    let prev2 = 1;  // ways to reach step i-2
    let prev1 = 2;  // ways to reach step i-1

    for (let i = 3; i <= n; i++) {
        const current = prev1 + prev2;
        prev2 = prev1;
        prev1 = current;
    }

    return prev1;
}

/**
 * Test cases for Problem 070: Climbing Stairs
 */
function testSolution() {
    console.log('Testing 070. Climbing Stairs');

    // Test case 1: n = 2
    const result1 = solve(2);
    const expected1 = 2;
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: n = 3
    const result2 = solve(3);
    const expected2 = 3;
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: n = 5
    const result3 = solve(5);
    const expected3 = 8;
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);

    console.log('All test cases passed for 070. Climbing Stairs!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 070. Climbing Stairs ===');
    console.log('Category: Dynamic Programming');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on dynamic programming concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */
